testcases/kernel/syscalls/write/write04.c

```diff
@@ -51,152 +51,126 @@
 #include <setjmp.h>
 #include <errno.h>
 #include <string.h>
-#include "test.h"
+#include <stdio.h>
+#include "tst_test.h"
 
 #define PIPE_SIZE_TEST getpagesize()
 
 void alarm_handler();
-void setup();
-void cleanup();
-
-char *TCID = "write04";
-int TST_TOTAL = 1;
 
 char fifo[100] = "fifo";
 static sigjmp_buf jmp;
 int rfd, wfd;
 
-int main(int argc, char **argv)
+static void verify_write(void)
 {
-	int lc;
-
 	struct stat buf;
 	int fail;
 	int cnt;
 	char wbuf[17 * PIPE_SIZE_TEST];
 	struct sigaction sigptr;	/* set up signal handler */
 
-	tst_parse_opts(argc, argv, NULL, NULL);
+	if (mknod(fifo, S_IFIFO | 0777, 0) < 0)
+		tst_res(TBROK | TERRNO, "mknod() failed, errno: %d", errno);
 
-	/* global setup */
-	setup();
+	if (stat(fifo, &buf) != 0)
+		tst_res(TBROK | TERRNO, "stat() failed, errno: %d", errno);
 
-	/*
-	 * The following loop checks looping state if -i option given
-	 */
-	for (lc = 0; TEST_LOOPING(lc); lc++) {
-		/* reset tst_count in case we are looping */
-		tst_count = 0;
+	if ((buf.st_mode & S_IFIFO) == 0)
+		tst_res(TBROK | TERRNO, "Mode does not indicate fifo file");
 
-		if (mknod(fifo, S_IFIFO | 0777, 0) < 0) {
-			tst_resm(TBROK, "mknod() failed, errno: %d", errno);
-			cleanup();
-		}
-		if (stat(fifo, &buf) != 0) {
-			tst_resm(TBROK, "stat() failed, errno: %d", errno);
-			cleanup();
-		}
-		if ((buf.st_mode & S_IFIFO) == 0) {
-			tst_resm(TBROK, "Mode does not indicate fifo file");
-			cleanup();
-		}
 #if 0
-		sigset(SIGALRM, alarm_handler);
+	sigset(SIGALRM, alarm_handler);
 #endif
-		sigptr.sa_handler = (void (*)(int signal))alarm_handler;
-		sigfillset(&sigptr.sa_mask);
-		sigptr.sa_flags = 0;
-		sigaddset(&sigptr.sa_mask, SIGALRM);
-		if (sigaction(SIGALRM, &sigptr, NULL) == -1) {
-			tst_resm(TBROK, "sigaction(): Failed");
-			cleanup();
-		}
+	sigptr.sa_handler = (void (*)(int signal))alarm_handler;
+	sigfillset(&sigptr.sa_mask);
+	sigptr.sa_flags = 0;
+	sigaddset(&sigptr.sa_mask, SIGALRM);
+	if (sigaction(SIGALRM, &sigptr, NULL) == -1)
+		tst_res(TBROK | TERRNO, "sigaction(): Failed");
+
 //block1:
-		tst_resm(TINFO, "Enter block 1: test for EAGAIN in write()");
-		fail = 0;
+	tst_res(TINFO, "Enter block 1: test for EAGAIN in write()");
+	fail = 0;
 
-		(void)memset((void *)wbuf, 'A', 17 * PIPE_SIZE_TEST);
+	(void)memset((void *)wbuf, 'A', 17 * PIPE_SIZE_TEST);
 
-		/*
-		 * open the read end of the pipe
-		 */
-		if (sigsetjmp(jmp, 1)) {
-			tst_resm(TBROK, "Error reading fifo, read blocked");
-			fail = 1;
-		}
-		(void)alarm(10);	/* set alarm for 10 seconds */
-		rfd = open(fifo, O_RDONLY | O_NONBLOCK);
-		(void)alarm(0);
-		if (rfd < 0) {
-			tst_resm(TBROK, "open() for reading the pipe failed");
-			fail = 1;
-		}
+	/*
+	 * open the read end of the pipe
+	 */
+	if (sigsetjmp(jmp, 1)) {
+		tst_res(TBROK | TERRNO, "Error reading fifo, read blocked");
+		fail = 1;
+	}
+	(void)alarm(10);	/* set alarm for 10 seconds */
+	rfd = open(fifo, O_RDONLY | O_NONBLOCK);
+	(void)alarm(0);
+	if (rfd < 0) {
```

Comment:
And also get rid of all the alarm() calls and the alarm handler and the sigsetjmp() mess, the new test library handles the test timeout itself, no need to add hacking timeouts like this one.